### PR TITLE
Avoid deep recursion while emitting deeply nested logical conditions

### LINKED ISSUE #66900: flow into dotnet/runtime main blocked
Runtime hasn't taken a new roslyn build for several weeks, the failure described here is blocking code flow.

**Version Used**: 

The roslyn flow into dotnet/runtime is breaking on https://github.com/dotnet/runtime/pull/81164 

**Steps to Reproduce**:

See linked issue

**Diagnostic Id**:

`src/libraries/System.Text.Json/tests/Common/TestClasses/TestClasses.Constructor.cs(2394,9): error CS8078: (NETCORE_ENGINEERING_TELEMETRY=Build) An expression is too long or complex to compile``

**Expected Behavior**:

[See the discussion](https://github.com/dotnet/runtime/pull/81164#issuecomment-1428475256)

**Actual Behavior**:

CS8078 when compiling a test that uses a 1000 parameter constructor

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #66900.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

**Comment by lewing** [comment]
cc @jaredpar

File: src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs [hunk 1]
```diff
@@ -415,43 +416,84 @@ private void EmitCondBranchCore(BoundExpression condition, ref object dest, bool
...
-
-                        case BinaryOperatorKind.LogicalAnd:
-                            if (testBothArgs)
-                            {
-                                // gotoif(a != sense) fallThrough
-                                // gotoif(b == sense) dest
-                                // fallThrough:
-
-                                object fallThrough = null;
+                        var stack = ArrayBuilder<(BoundExpression? condition, StrongBox<object?> destBox, bool sense)>.GetInstance();
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs:426]: > Why is dest wrapped in a ```StrongBox```? Why not store just ```object?``` in the stack?

Its value is mutable and is supposed to be shared between  different items in the stack (including mutations). If we simply store ```object?```, we will store its current value (often `null`) and won't be able to observe the changes. We need an indirection in order to be able share the mutations as well. ```StrongBox``` gives us an indirection.

**Review by jjonescz — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs [hunk 2]
```diff
@@ -415,43 +416,84 @@ private void EmitCondBranchCore(BoundExpression condition, ref object dest, bool
...
+                                    // gotoif(b == sense) labDest
 
-                                EmitCondBranch(binOp.Left, ref dest, sense);
-                                condition = binOp.Right;
+                                    // Note, operations are pushed to the stack in opposite order
+                                    stack.Push((binary.Right, top.destBox, top.sense));
+                                    stack.Push((binary.Left, top.destBox, top.sense));
+                                }
+                            }
+                            else if (stack.Count == 0 && ReferenceEquals(destBox.Value, top.destBox.Value))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs:470]:
```suggestion
                            else if (stack.Count == 0 && ReferenceEquals(destBox, top.destBox))
```

**Review by cston — APPROVED** [review]

**Review by jjonescz — APPROVED** [review]

**Pull request merged**